### PR TITLE
Implement migration.json support when importing

### PR DESCRIPTION
**Pull request by adrianchivu**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/CDP4-SDK-Community-Edition/pulls) open
- [x] I have verified that I am following the CDP4-SDK [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/CDP4-SDK-Community-Edition/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### CODE REVIEW

**Review by alexatrhea — APPROVED** [review]

**Review by samatrhea — APPROVED** [review]

**Comment by cozminvelciu** [comment]
Is this reviewable? I see it still has the WIP tag.

**Comment by adrianchivu** [comment]
> Is this reviewable? I see it still has the WIP tag.

Ready for review.

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 1]
```diff
@@ -243,6 +248,8 @@ public override Task<IEnumerable<Thing>> Write(IEnumerable<OperationContainer> o
                     this.WriteModelReferenceDataLibraryToZipFile(modelReferenceDataLibraries, zipFile, path);
 
                     this.WriteIterationsToZipFile(iterations, zipFile, path);
+
+                    this.WriteExtraFilesToZipFile(files, zipFile, path);
```
Comment (cozminvelciu) [inline CDP4JsonFileDal/JsonFileDal.cs:252]: I would rename the "files" variable to reflect what it actually signifies: extensionFiles.

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 2]
```diff
@@ -806,6 +813,28 @@ private void WriteIterationsToZipFile(IEnumerable<CDP4Common.EngineeringModelDat
             }
         }
 
+        /// <summary>
+        ///  Writes the application dependend files inside specific folder to the <see cref="ZipFile"/>
+        /// </summary>
+        /// <param name="extraFilesPath">The files list that will be written</param>
+        /// <param name="zipFile">The target <see cref="ZipFile"/></param>
+        /// <param name="filePath">The file of the target <see cref="ZipFile"/></param>
```
Comment (cozminvelciu) [inline CDP4JsonFileDal/JsonFileDal.cs:821]: "The file of" -> "The file path of"

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 3]
```diff
@@ -806,6 +813,28 @@ private void WriteIterationsToZipFile(IEnumerable<CDP4Common.EngineeringModelDat
             }
         }
 
+        /// <summary>
+        ///  Writes the application dependend files inside specific folder to the <see cref="ZipFile"/>
+        /// </summary>
+        /// <param name="extraFilesPath">The files list that will be written</param>
+        /// <param name="zipFile">The target <see cref="ZipFile"/></param>
+        /// <param name="filePath">The file of the target <see cref="ZipFile"/></param>
+        private void WriteExtraFilesToZipFile(IEnumerable<string> extraFilesPath, ZipFile zipFile, string filePath)
```
Comment (cozminvelciu) [inline CDP4JsonFileDal/JsonFileDal.cs:822]: I would rename "WriteExtra" -> "WriteExtension"

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 4]
```diff
@@ -806,6 +813,28 @@ private void WriteIterationsToZipFile(IEnumerable<CDP4Common.EngineeringModelDat
...
+        {
+            if (extraFilesPath is null)
+            {
+                return;
+            }
+
+            foreach (var extraFile in extraFilesPath)
+            {
+                var zipEntry = zipFile.AddFile(extraFile, ExtensionsZipLocation);
+                zipEntry.Comment = $"The {extraFile} file";
```
Comment (cozminvelciu) [inline CDP4JsonFileDal/JsonFileDal.cs:832]: There's no information (value) added by this comment, so I propose we remove it.

File: CDP4JsonFileDal/JsonFileDal.cs [hunk 5]
```diff
@@ -80,6 +80,11 @@ public class JsonFileDal : Dal
         /// </summary>
         private const string EngineeringModelZipLocation = "EngineeringModels";
 
+        /// <summary>
+        /// The application-dependent(extensions) files zip location.
```
Comment (cozminvelciu) [inline CDP4JsonFileDal/JsonFileDal.cs:84]: Add whitespace before punctuation marks (parentheses).

File: CDP4JsonFileDal.Tests/files/migration.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    "1c2466ed-ddb9-4546-b458-5b3dbcb58924": {
+      "password": "70d3491b7b5033ce7e4f77630058a2b78a8e5a817b7eb957a306490092d4dd83",
+      "salt": "8b747da1d2220d2deb9c153882177727edcf05151b9c575881eea57fce7ecceb"
+    },
+    "cc89eec4-d7ba-486d-854b-8f2407e154e2": {
+      "password": "70d3491b7b5033ce7e4f77630058a2b78a8e5a817b7eb957a306490092d4dd83",
+      "salt": "8b747da1d2220d2deb9c153882177727edcf05151b9c575881eea57fce7ecceb"
+    }
+  }
+}
```
Comment (cozminvelciu) [inline CDP4JsonFileDal.Tests/files/migration.json:13]: Missing newline at end of file.

File: CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs [hunk 7]
```diff
@@ -99,12 +109,23 @@ public void OneTimeSetUp()
         public void SetUp()
         {
             var path = Path.Combine(TestContext.CurrentContext.TestDirectory, "files", "LOFT_ECSS-E-TM-10-25_AnnexC.zip");
+            var migrationSourceFile = Path.Combine(TestContext.CurrentContext.TestDirectory, @"..\..\..\files", "migration.json");
```
Comment (cozminvelciu) [inline CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs:112]: I might not be understanding this correctly, but does this file path point outside of the project?

File: CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs [hunk 8]
```diff
@@ -99,12 +109,23 @@ public void OneTimeSetUp()
         public void SetUp()
         {
             var path = Path.Combine(TestContext.CurrentContext.TestDirectory, "files", "LOFT_ECSS-E-TM-10-25_AnnexC.zip");
+            var migrationSourceFile = Path.Combine(TestContext.CurrentContext.TestDirectory, @"..\..\..\files", "migration.json");
+
+            this.annexC3File = Path.Combine(TestContext.CurrentContext.TestDirectory, "files", "AnnexC3.zip");
+            this.migrationFile = Path.Combine(TestContext.CurrentContext.TestDirectory, "files", "migration.json");
+
+            if (!File.Exists(this.migrationFile))
```
Comment (cozminvelciu) [inline CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs:117]: Does this try to copy the test migration file from outside the project, if it does not exist?

**Review by cozminvelciu — CHANGES_REQUESTED** [review]

File: CDP4JsonFileDal/JsonFileDal.cs (hunk 1 above)
Comment (samatrhea) [reply CDP4JsonFileDal/JsonFileDal.cs:252]: agreed

File: CDP4JsonFileDal/JsonFileDal.cs (hunk 3 above)
Comment (samatrhea) [reply CDP4JsonFileDal/JsonFileDal.cs:822]: i agree

File: CDP4JsonFileDal/JsonFileDal.cs (hunk 4 above)
Comment (adrianchivu) [reply CDP4JsonFileDal/JsonFileDal.cs:832]: All archive components have this property set so I propose to keep it as it is for the moment.

File: CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs (hunk 7 above)
Comment (adrianchivu) [reply CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs:112]: TestContext.CurrentContext.TestDirectory point to test project source path /bin/$(SolutionConfiguration)/${TargetFramework} and files path its local path inside test project.

File: CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs (hunk 8 above)
Comment (adrianchivu) [reply CDP4JsonFileDal.Tests/JsonFileDalTestFixture.cs:117]: See above comment.

**Review by alexatrhea — APPROVED** [review]

**Review by samatrhea — APPROVED** [review]

**Pull request merged**